include/boost/random/niederreiter_base2.hpp

```diff
@@ -0,0 +1,371 @@
+inline typename boost::dynamic_bitset<Block, Allocator>::size_type
+bitset_log2(const boost::dynamic_bitset<Block, Allocator>& v)
+{
+  typedef boost::dynamic_bitset<Block, Allocator> bitset_t;
+  typedef typename bitset_t::size_type size_type;
+
+  if (v.none())
+    throw std::invalid_argument("bitset_log2");
+
+  size_type up  = v.size() - 1;
+  size_type low = v.find_next(0);
+
+  // Binary lookup for the most significant set bit
+  while (low < up)
+  {
+    size_type m = low + (up - low) / 2;
+
+    // Check if any bit is present after mid
+    size_type p = v.find_next(m);
+    if (p != bitset_t::npos)
+        low = p;
+    else
+        up = m;
+  }
+
+  return low;
+}
+
+
+// Multiply polynomials over Z_2.
+template <typename Block, typename Allocator>
+inline boost::dynamic_bitset<Block, Allocator>
+modulo2_multiply(unsigned P, boost::dynamic_bitset<Block, Allocator> v)
+{
+  boost::dynamic_bitset<Block, Allocator> pt (v.size());
+  for (; P; P >>= 1, v <<= 1)
+    if (P & 1) pt ^= v;
+  return pt;
+}
+
+
+// Calculate the values of the constants V(J,R) as
+// described in BFN section 3.3.
+//
+// px = appropriate irreducible polynomial for current dimension
+// pb = polynomial defined in section 2.3 of BFN.
+// pb is modified
+template <typename Block, typename Allocator, typename T>
+inline void calculate_v(const boost::dynamic_bitset<Block, Allocator>& pb,
+  typename boost::dynamic_bitset<Block, Allocator>::size_type& pb_degree,
+  std::vector<T>& v)
+{
+  typedef typename boost::dynamic_bitset<Block, Allocator>::size_type size_type;
+
+  const T arbitrary_element = static_cast<T>(1);  // arbitray element of Z_2
+
+  // Now choose a value of Kj as defined in section 3.3.
+  // We must have 0 <= Kj < E*J = M.
+  // The limit condition on Kj does not seem very relevant
+  // in this program.
+  size_type kj = pb_degree;
+
+  pb_degree = bitset_log2(pb);
+
+  // Now choose values of V in accordance with
+  // the conditions in section 3.3.
+  std::fill(v.begin(), v.begin() + kj, T());
+
+  // Quoting from BFN: "Our program currently sets each K_q
+  // equal to eq. This has the effect of setting all unrestricted
+  // values of v to 1."
+  // Actually, it sets them to the arbitrary chosen value.
+  // Whatever.
+  for (size_type r = kj; r < pb_degree; ++r)
+    v[r] = arbitrary_element;
+
+  // Calculate the remaining V's using the recursion of section 2.3,
+  // remembering that the B's have the opposite sign.
+  for (size_type r = pb_degree; r < v.size(); ++r)
+  {
+    T term = T /*zero*/ ();
+    boost::dynamic_bitset<Block, Allocator> pb_c = pb;
+    for (int k = -pb_degree; k < 0; ++k, pb_c >>= 1)
+    {
+      if( pb_c.test(0) )
+        term ^= v[r + k];
+    }
+    v[r] = term;
+  }
+}
+
+} // namespace nb2
+
+template<typename IntType, typename Nb2Table>
+struct niederreiter_base2_lattice
+{
+  typedef IntType value_type;
+
+  BOOST_STATIC_CONSTANT(unsigned, bit_count = std::numeric_limits<value_type>::digits);
```

Comment:
This should be a template parameter, rather than being deduced from the size of the argument.  The goal is to make it so that the output of the generator is platform independent.